src/ray/raylet/worker_pool.cc

```diff
@@ -1224,122 +1191,87 @@ void WorkerPool::PopWorker(const TaskSpecification &task_spec,
-                start_worker_process_fn(task_spec,
-                                        state,
-                                        dynamic_options,
-                                        true,
-                                        serialized_runtime_env_context,
-                                        callback);
-              } else {
-                process_failed_runtime_env_setup_failed_++;
-                callback(nullptr,
-                         PopWorkerStatus::RuntimeEnvCreationFailed,
-                         /*runtime_env_setup_error_message*/ setup_error_message);
-                RAY_LOG(WARNING)
-                    << "Create runtime env failed for task " << task_spec.TaskId()
-                    << " and couldn't create the dedicated worker.";
-              }
-            },
-            allocated_instances_serialized_json);
-      } else {
-        start_worker_process_fn(task_spec, state, dynamic_options, true, "", callback);
-      }
+  }
+
+  bool is_actor_creation = task_spec.IsActorCreationTask();
+  std::vector<std::string> dynamic_options{};
+  if (is_actor_creation && !task_spec.DynamicWorkerOptions().empty()) {
+    dynamic_options = task_spec.DynamicWorkerOptions();
+  }
+
+  const int runtime_env_hash = task_spec.GetRuntimeEnvHash();
+  for (auto it = idle_of_all_languages_.rbegin(); it != idle_of_all_languages_.rend();
+       it++) {
+    if (task_spec.GetLanguage() != it->first->GetLanguage() ||
+        it->first->GetAssignedJobId() != task_spec.JobId() ||
+        state.pending_disconnection_workers.count(it->first) > 0 || it->first->IsDead()) {
+      continue;
     }
-  } else {
-    // Find an available worker which is already assigned to this job and which has
-    // the specified runtime env.
-    // Try to pop the most recently pushed worker.
-    const int runtime_env_hash = task_spec.GetRuntimeEnvHash();
-    for (auto it = idle_of_all_languages_.rbegin(); it != idle_of_all_languages_.rend();
-         it++) {
-      if (task_spec.GetLanguage() != it->first->GetLanguage() ||
-          it->first->GetAssignedJobId() != task_spec.JobId() ||
-          state.pending_disconnection_workers.count(it->first) > 0 ||
-          it->first->IsDead()) {
-        continue;
-      }
-      // These workers are exiting. So skip them.
-      if (pending_exit_idle_workers_.count(it->first->WorkerId())) {
-        continue;
-      }
-      // Skip if the runtime env doesn't match.
-      if (runtime_env_hash != it->first->GetRuntimeEnvHash()) {
-        continue;
-      }
 
-      state.idle.erase(it->first);
-      // We can't erase a reverse_iterator.
-      auto lit = it.base();
-      lit--;
-      worker = std::move(lit->first);
-      idle_of_all_languages_.erase(lit);
-      idle_of_all_languages_map_.erase(worker);
-      break;
+    // Skip if the dynamic_options doesn't match.
+    if (LookupWorkerDynamicOptions(it->first->GetStartupToken()) != dynamic_options) {
```

Comment:
Should we index this? This could be a bit expensive to iterate all workers for each iteration? (It's N^2?)

Reply:
ah actually most of time it's O(N). The `LookupWorkerDynamicOptions` is only O(1).